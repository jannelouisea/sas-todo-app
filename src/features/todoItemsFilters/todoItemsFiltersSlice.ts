import { createSlice, PayloadAction } from '@reduxjs/toolkit'

import { Sort } from 'src/static/enums'
import { ITodoItemsFilters } from 'src/interfaces';

const initialState: ITodoItemsFilters = {
    searchTerm: '',
    showNewer: true,
    showOlder: true,
    showCompleted: true,
    sortBy: Sort.Incmpl1st,
};

const todoItemsFiltersSlice = createSlice({
    name: 'todoItemsFilters',
    initialState,
    reducers: {
        searchTermChanged(state, action: PayloadAction<string>) {
            return { ...state, searchTerm: action.payload }
        },
        showNewerToggled(state) {
            return { ...state, showNewer: !state.showNewer }
        },
        showOlderToggled(state) {
            return { ...state, showOlder: !state.showOlder }
        },
        showCompletedToggled(state) {
            return { ...state, showCompleted: !state.showCompleted }
        },
        sortByChanged(state, action: PayloadAction<Sort>) {
            return { ...state, sortBy: action.payload }
        }
    }
});

export const {
    searchTermChanged,
    showNewerToggled,
    showOlderToggled,
    showCompletedToggled,
    sortByChanged,
} = todoItemsFiltersSlice.actions;

export default todoItemsFiltersSlice.reducer;