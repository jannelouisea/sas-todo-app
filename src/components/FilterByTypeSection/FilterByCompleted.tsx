import {
    FilterByButton
} from 'src/components'
import DoneIcon from '@mui/icons-material/Done';
import PropTypes from 'prop-types';
import {
    MUIColor
} from 'src/static/enums'

import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { showCompletedToggled } from 'src/features/todoItemsFilters/todoItemsFiltersSlice';

type Props = {
    className: string
}

function FilterByCompleted({ className }: Props) {
    const showCompleted = useAppSelector(state => state.todoItemsFilters.showCompleted);
    const dispatch = useAppDispatch();

    const onToggleFilter = () => {
        dispatch(showCompletedToggled());
    }

    return <FilterByButton
        className={className}
        label='Completed'
        color={MUIColor.Success}
        startIcon={<DoneIcon fontSize='small' />}
        filterEnabled={showCompleted}
        onClick={onToggleFilter}
    />;
}

FilterByCompleted.propTypes = {
    className: PropTypes.string,
}

FilterByCompleted.defaultProps = {
    className: ''
}

export default FilterByCompleted;