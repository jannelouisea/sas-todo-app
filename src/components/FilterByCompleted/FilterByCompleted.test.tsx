import FilterByCompleted from './FilterByCompleted';
import { Provider } from 'react-redux';
import {
    IStoreType,
    createStore
} from 'src/app/store';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import * as todoItemsFiltersActions from 'src/features/todoItemsFilters/todoItemsFiltersSlice';

describe('FilterByCompleted', () => {
    let store: IStoreType;

    beforeEach(() => {
        store = createStore();
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <FilterByCompleted className='foo' />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('toggles completed filter on click', () => {
        const todoItemCompletedSpy = jest.spyOn(todoItemsFiltersActions, 'showCompletedToggled');

        render(
            <Provider store={store}>
                <FilterByCompleted className='foo' />
            </Provider>
        );

        let button: HTMLInputElement = screen.getByLabelText('Filter by completed');
        expect(button).toBeInTheDocument();
        fireEvent.click(button);

        expect(todoItemCompletedSpy).toHaveBeenCalledTimes(1);
    })
});