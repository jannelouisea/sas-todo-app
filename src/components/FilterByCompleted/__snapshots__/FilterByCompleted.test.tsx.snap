// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterByCompleted renders correctly 1`] = `
<div
  className="foo"
>
  <button
    aria-label="Filter by completed"
    className="MuiButton-root MuiButton-variantSolid MuiButton-colorSuccess MuiButton-sizeSm foo css-1yqcigs-JoyButton-root"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="MuiButton-startDecorator css-187ulk2-JoyButton-startDecorator"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="DoneIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
        />
      </svg>
    </span>
    Completed
  </button>
</div>
`;
